provider/otc/otc.go

```diff
@@ -0,0 +1,209 @@
+package otc
+
+import (
+	"bytes"
+	"crypto/tls"
+	"encoding/json"
+	"fmt"
+	"github.com/messagebird/sachet"
+	"io"
+	"io/ioutil"
+	"net/http"
+	"time"
+)
+
+type OTCConfig struct {
+	IdentityEndpoint string `yaml:"identity_endpoint"`
+	DomainName       string `yaml:"domain_name"`
+	ProjectName      string `yaml:"project_name"`
+	UserName         string `yaml:"username"`
+	Password         string `yaml:"password"`
+	ProjectID        string `yaml:"project_id"`
+	Insecure         bool   `yaml:"insecure"`
+	token            string
+	otcBaseURL       string
+}
+
+type OTC struct {
+	OTCConfig
+}
+
+func NewOTC(config OTCConfig) *OTC {
+	OTC := &OTC{config}
+	return OTC
+}
+
+func (c *OTC) loginRequest() error {
+	type nameResponse struct {
+		Name string `json:"name"`
+	}
+
+	type userResponse struct {
+		Name     string       `json:"name"`
+		Password string       `json:"password"`
+		Domain   nameResponse `json:"domain"`
+	}
+
+	type passwordResponse struct {
+		User userResponse `json:"user"`
+	}
+	type identityResponse struct {
+		Methods  []string         `json:"methods"`
+		Password passwordResponse `json:"password"`
+	}
+
+	type scopeResponse struct {
+		Project nameResponse `json:"project"`
+	}
+
+	type authResponse struct {
+		Identity identityResponse `json:"identity"`
+		Scope    scopeResponse    `json:"scope"`
+	}
+
+	type loginResponse struct {
+		Auth authResponse `json:"auth"`
+	}
+
+	userResp := userResponse{
+		Name:     c.UserName,
+		Password: c.Password,
+		Domain: nameResponse{
+			Name: c.DomainName,
+		},
+	}
+
+	loginResp := loginResponse{
+		Auth: authResponse{
+			Identity: identityResponse{
+				Methods: []string{"password"},
+				Password: passwordResponse{
+					User: userResp,
+				},
+			},
+			Scope: scopeResponse{
+				Project: nameResponse{
+					Name: c.ProjectName,
+				},
+			},
+		},
+	}
+
+	body, err := json.Marshal(loginResp)
+	if err != nil {
+		return err
+	}
+	req, err := http.NewRequest("POST", c.IdentityEndpoint, bytes.NewReader(body))
+	if err != nil {
+		return err
+	}
+	req.Header.Set("Content-Type", "application/json")
+
+	tr := http.DefaultTransport.(*http.Transport)
+	tr.TLSClientConfig = &tls.Config{InsecureSkipVerify: c.Insecure}
+
+	client := &http.Client{Timeout: time.Duration(10 * time.Second), Transport: tr}
+	resp, err := client.Do(req)
+	if err != nil {
+		return err
+	}
+	defer resp.Body.Close()
+
+	if resp.StatusCode >= 400 {
+		return fmt.Errorf("OTC API request failed with HTTP status code %d", resp.StatusCode)
+	}
+
+	c.token = resp.Header.Get("X-Subject-Token")
+
+	if c.token == "" {
+		return fmt.Errorf("unable to get auth token")
+	}
+
+	type endpointResponse struct {
+		Token struct {
+			Catalog []struct {
+				Type      string `json:"type"`
+				Endpoints []struct {
+					URL string `json:"url"`
+				} `json:"endpoints"`
+			} `json:"catalog"`
+		} `json:"token"`
+	}
+	var endpointResp endpointResponse
+
+	err = json.NewDecoder(resp.Body).Decode(&endpointResp)
```

Comment:
why do you parse the response into `endpointResponse` if you are not going to use it?